rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/wait.hpp

```diff
@@ -15,15 +15,103 @@
 #ifndef RMW_CONNEXT_SHARED_CPP__WAIT_HPP_
 #define RMW_CONNEXT_SHARED_CPP__WAIT_HPP_
 
+#include <unordered_map>
+#include <unordered_set>
+
 #include "ndds_include.hpp"
 
 #include "rmw/error_handling.h"
 #include "rmw/impl/cpp/macros.hpp"
 #include "rmw/types.h"
 
 #include "rmw_connext_shared_cpp/condition_error.hpp"
+#include "rmw_connext_shared_cpp/event_converter.hpp"
 #include "rmw_connext_shared_cpp/types.hpp"
 #include "rmw_connext_shared_cpp/visibility_control.h"
+#include "rmw_connext_shared_cpp/connext_static_event_info.hpp"
+
+
+rmw_ret_t __gather_event_conditions(
+  rmw_events_t * events,
+  std::unordered_set<DDS::StatusCondition *> & status_conditions)
+{
+  RMW_CHECK_ARGUMENT_FOR_NULL(events, RMW_RET_INVALID_ARGUMENT);
+
+  std::unordered_map<DDS::StatusCondition *, DDS::StatusMask> status_mask_map;
+  // gather all status conditions and masks
+  for (size_t i = 0; i < events->event_count; ++i) {
+    rmw_event_t * current_event = static_cast<rmw_event_t *>(events->events[i]);
+    DDS::Entity * dds_entity = static_cast<ConnextCustomEventInfo *>(
+      current_event->data)->get_entity();
+    if (!dds_entity) {
+      RMW_SET_ERROR_MSG("Event handle is null");
+      return RMW_RET_ERROR;
+    }
+    DDS::StatusCondition * status_condition = dds_entity->get_statuscondition();
+    if (!status_condition) {
+      RMW_SET_ERROR_MSG("status condition handle is null");
+      return RMW_RET_ERROR;
+    }
+    if (is_event_supported(current_event->event_type)) {
```

Comment:
What if it isn't? Should we continue? Should there be an error raised?

Reply:
If this event is not supported it would affect the other conditions if we didn't continue (so it's a no-op for the unsupported event). We haven't been setting the unsupported message error, but I'll add here and in a few other places.